### PR TITLE
Implement witgen for intermediate polynomials

### PR DESCRIPTION
**Pull request by Schaeff**
This is done by inlining the intermediate polynomials at the beginning of witgen.

Todo:
- [ ] more tests

### CODE REVIEW

**Comment by Schaeff** [comment]
Yeah that's what I meant by more tests, not sure the visitor works on nested

**Comment by leonardoalt** [comment]
yea it was wrong, it needs to be post-visit

File: executor/src/witgen/util.rs [hunk 1]
```diff
@@ -66,3 +69,71 @@ pub fn substitute_constants<T: Copy>(
...
+                }
+            }
+        }
+        std::ops::ControlFlow::Continue::<()>(())
+    });
+    expr
+}
+
+pub fn substitute_intermediate<T: Copy>(
+    identities: &[Identity<T>],
```
Comment (Schaeff) [inline executor/src/witgen/util.rs:107]: This could be passed as value to avoid cloning inside this function

File: executor/src/witgen/util.rs [hunk 2]
```diff
@@ -66,3 +69,71 @@ pub fn substitute_constants<T: Copy>(
         })
         .collect()
 }
+
+fn inline_intermediates_in_expression<T: Copy>(
+    expr: &Expression<T>,
```
Comment (Schaeff) [inline executor/src/witgen/util.rs:74]: This could be passed as value to avoid cloning inside this function

**Comment by Schaeff** [comment]
Need to do the same in halo2 and we're done

**Comment by leonardoalt** [comment]
Implemented inlining before halo2 and squashed

**Comment by leonardoalt** [comment]
@Schaeff I moved the inlining stuff into pil_analyzer now so that it can be used by both witgen and halo2 directly. pil_analyzer was already a dependency of these 2

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**